### PR TITLE
Improve lyrics lookup by removing remix & remaster info in query.

### LINKED ISSUE #250: Genius API is sometimes weird
**Is your feature request related to a problem? Please describe.**
Getting lyrics from its API behaves a bit weird sometimes.

Searching in their api "Peace Sells Megadeth" show the correct song, but "Peace Sells Remastered Megadeth" does not! And it even does not show Megadeth after the first results.

**Describe the solution you'd like**
Don't know how this could be solved.

**Describe alternatives you've considered**
Maybe we could search for a different API service which we could pass the song and the author as parameters, unlike Genius which is a general term.
Or maybe we could ask the user "is this lyrics wrong" yes -> "please type a search"?

I don't think this has a solution in the very near future, but tracking it could be useful, that's why I created this issue.

By the way, awesome project!! I'm enjoying a lot using it :)

### PR DESCRIPTION
**Pull request by Icelk**
Fixes #250 .

Just simple string manipulation to remove remaster & remix info from query.

I've tried it out and it fixes all my weird lyrics.

### CODE REVIEW

File: lyric_finder/src/lib.rs [hunk 1]
```diff
@@ -136,6 +138,45 @@ impl Default for Client {
...
 
+fn improve_query(query: &str) -> String {
+    let is_dash = |c: char| c == '-' || c == '—' || c == '–';
+    let r_no_filler = |s: &str, idx: usize| {
+        let Some(s) = s.get(..idx) else { return idx };
+        s.char_indices()
+            .rev()
+            .find(|(_, c)| !(is_dash(*c) || c.is_whitespace()))
+            .map_or(idx, |(idx, c)| idx + c.len_utf8())
+    };
```
Comment (aome510) [inline lyric_finder/src/lib.rs:149]: what is `r_no_filter` for?

File: lyric_finder/src/lib.rs [hunk 2]
```diff
@@ -136,6 +138,45 @@ impl Default for Client {
     }
 }
 
+fn improve_query(query: &str) -> String {
```
Comment (aome510) [inline lyric_finder/src/lib.rs:141]: can you add a function document explaining the function's purpose? Maybe, "Improve the query by filtering out redundant parts, e.g `Remix`, `Remsatered` that may result in incorrect search results"?

File: lyric_finder/src/lib.rs [hunk 3]
```diff
@@ -136,6 +138,45 @@ impl Default for Client {
...
+        let Some(s) = s.get(..idx) else { return idx };
+        s.char_indices()
+            .rev()
+            .find(|(_, c)| !(is_dash(*c) || c.is_whitespace()))
+            .map_or(idx, |(idx, c)| idx + c.len_utf8())
+    };
+
+    let mut query = query.to_owned();
+    // just cut xxxx Remaster from query
+    if let Some(remaster_start) = query.find("Remaster").or_else(|| query.find("remaster")) {
```
Comment (aome510) [inline lyric_finder/src/lib.rs:153]: can process the query by `query = query.lower_bound()` so you don't need to manually handle cases like this

**Review by aome510 — CHANGES_REQUESTED** [review]

File: lyric_finder/src/lib.rs (hunk 3 above)
Comment (Icelk) [reply lyric_finder/src/lib.rs]: Query is now always converted to lowercase. That shouldn't matter for search results.

File: lyric_finder/src/lib.rs (hunk 1 above)
Comment (Icelk) [reply lyric_finder/src/lib.rs]: Now documented in code:
```rust
// reverse finder for non-filler (space, dashes) chars.
// Acts like a trim to remove undesired spaces and dashes.
```

File: lyric_finder/src/lib.rs [hunk 4]
```diff
@@ -136,6 +138,60 @@ impl Default for Client {
     }
 }
 
+/// Returns `query` without `remaster` & `remix` information from track/artist query.
+/// Returned value is lowercase.
+/// These caused wildly invalid lyrics to be found.
+/// (try yourself adding remastered 2011 to a song's name when searching in Genius!)
+fn improve_query(query: &str) -> String {
+    let is_dash = |c: char| c == '-' || c == '—' || c == '–';
```
Comment (aome510) [inline lyric_finder/src/lib.rs:146]: this is interesting, doesn't Spotify only use `-` as the dash in most songs? Looking into the codes, it's hard to see the differences between these three cases. IMO, it should be fine to check with `-` only.

File: lyric_finder/src/lib.rs [hunk 5]
```diff
@@ -136,6 +138,60 @@ impl Default for Client {
...
+        trim_start = r_no_filler(&query, trim_start);
+        let end = end_of_word(&query, end);
+        query.drain(trim_start..end);
+    }
+    // just cut anything off after - for remix
+    if let Some(remix_start) = query.find("remix") {
+        let end = remix_start + "remix".len();
+        let end = end_of_word(&query, end);
+        if let Some(metadata_start) = query.rfind(is_dash) {
+            if metadata_start > 4 {
```
Comment (aome510) [inline lyric_finder/src/lib.rs:186]: where does the `4` come from?

File: lyric_finder/src/lib.rs [hunk 6]
```diff
@@ -136,6 +138,60 @@ impl Default for Client {
...
+    };
+    // used to handle longer variants of words: `remixed`, `remastered`, etc.
+    let end_of_word = |s: &str, idx: usize| {
+        let Some(s) = s.get(idx..) else { return idx };
+        s.find(|c| !c.is_alphanumeric())
+            .map_or(idx, |found| found + idx)
+    };
+
+    let mut query = query.to_lowercase();
+    // just cut xxxx Remaster from query
```
Comment (aome510) [inline lyric_finder/src/lib.rs:165]:
```suggestion
    // remove "xxxx Remaster" from query
    // For example, `{song} xxxx Remastered {artists}` becomes `{song} {artists}`.
```

File: lyric_finder/src/lib.rs [hunk 7]
```diff
@@ -136,6 +138,60 @@ impl Default for Client {
...
+        if let Some(year) = query.get(end_of_prev_word + 1..remaster_start.saturating_sub(1)) {
+            if year.chars().all(|c| c.is_whitespace() || c.is_numeric()) {
+                trim_start = end_of_prev_word;
+            }
+        }
+        trim_start = r_no_filler(&query, trim_start);
+        let end = end_of_word(&query, end);
+        query.drain(trim_start..end);
+    }
+    // just cut anything off after - for remix
```
Comment (aome510) [inline lyric_finder/src/lib.rs:181]:
```suggestion
    // remove "- xxxx remix" from the query
    // For example, `{song} - xxxx remix {artists}` becomes `{song} {artists}`.
```

File: lyric_finder/src/lib.rs (hunk 4 above)
Comment (Icelk) [reply lyric_finder/src/lib.rs:146]: Did it just in case. Will change to your preference :)

File: lyric_finder/src/lib.rs (hunk 5 above)
Comment (Icelk) [reply lyric_finder/src/lib.rs]: Documented in the latest commit:
```rust
    // we'll flag for doing something wrong if the new song name is too short.
    const ARBETRARY_SONG_MIN_LENGTH: usize = 3;

```

**Comment by Icelk** [comment]
CI should be working now too.

**Review by aome510 — APPROVED** [review]
LGTM! I did a bit of cleanup before merging.

Tbh, after thinking more about the approach, it will be better to handle the removal with regex instead of manually finding the word boundary, and characters or checking year format, which feels kinda hacky.

For the first iteration, I'm good with the approach in this PR.

**Comment by aome510** [comment]
Also, thanks @Icelk for the PR.

**Pull request merged**